Add weighted-round-robin policy to loadbalance plugin

gabordozsa opened the pull request:
Weighted round robin policy gives more control over the load distribution then random shuffle. Weight values assigned to IPs determine the relative likelihood of an IP being returned as top record in the answer.

Signed-off-by: Gabor Dozsa <gabor.dozsa@ibm.com>

<!--
Thank you for contributing to CoreDNS!
Please provide the following information to help us make the most of your pull request:
-->

### 1. Why is this pull request needed and what does it do?

This PR extends the **loadbalance** plugin with a new "weighted-round-robin" strategy.  

The loadbalance plugin only provides a random shuffle of records currently. Sometimes more control over the load distribution is required (e.g. hierarchical load balancing among non-uniform resources). Weighted-round-robin is a simple enough strategy where the weight value assigned to an address defines the relative likelihood of an IP being returned as top record in the answer.

### 2. Which issues (if any) are related?

https://github.com/coredns/coredns/issues/5616

### 3. Which documentation changes (if any) need to be made?

The PR proposes changes to the README of the **loadbalance** plugin.

### 4. Does this introduce a backward incompatible change or deprecation?
No

gabordozsa committed:
Add weighted-round-robin policy to loadbalance plugin

Weighted round robin policy gives more control over
the load distribution then random shuffle. Weight values
assigned to IPs determine the relative likelihood of
an IP being returned as top record in the answer.

Signed-off-by: Gabor Dozsa <gabor.dozsa@ibm.com>

gabordozsa commented on plugin/loadbalance/setup.go:
```diff
@@ -10,34 +14,105 @@ import (
 )
 
 var log = clog.NewWithPlugin("loadbalance")
+var openErr = errors.New("Weight file open error")
```
Ack

chrisohaver commented on plugin/loadbalance/README.md:
```diff
@@ -19,7 +19,40 @@ implementations (like glibc) are particular about that.
+
+The weight file is parsed line-by-line. If the first character in the line is '#' then the line is ignored (i.e. comment line). Otherwise, if the line contains a single word then it is a domain name. If the line contains two words then the first one is an IP and the second one is a weight value for that IP. The IPs are addresses for the domain name which is defined above them in the file. Multipl ...
+
+If the server receives a query for which no domain name is specified in the corresponding weight file then the answer is returned unmodified. If the domain name is specified in the weight file but the answer does not include the expected top IP (or there isn't any IP/weight pair specified for the particular domain name) then the answer is returned unmodified.
```
> If the domain name is specified in the weight file but the answer does not include the expected top IP ...

I don't understand what "the expected top IP" means here?   Do you mean "If the domain name is specified in the weight file but the list of IPs in the answer does not the same as the list of weighted IPs in the weight file for that domain ... "

gabordozsa replied:
Not really. The weighted-round-robin policy does not shuffle all the elements of the result lists (as opposed to the default round-robin). It only tries to move a particular IP to the first position in the list.  So, whenever it gets a new result list it does the following two steps: 

1.  it computes the "next expected IP" from the weighted IPs list 
2.  it checks if that particular IP is in the result list. If so, then it moves that IP to the first (top) position in the list.

The rest of the result list is not modified (i.e. not shuffled). The weights only determine how likely an IP will be returned as the first item in the result list.

chrisohaver replied:
Thanks.  IMO the wording in the readme doesn't clearly capture that.

The behavior seems counterintuitive - that results will be sometimes weighted and sometimes not for the same query name and answer set.

It would seem more intuitive to only return the answer unaltered if the answer contains a record that is not defined in the weight file.  Otherwise, the answer records all have defined weights, and one can be selected to be the top/first record in the response.

An alternate approach would be to only return the answer unaltered if the answer does not contain any records defined in the weight file.  Otherwise, only records with defined weights are eligible to be selected for the top/first record in the response.  Additionally, when returning the answer "unaltered", it would probably be better to randomly select a top IP, as if all the records have equal weight (which in this case they do in a sense).

Yet another approach would be to define in the weight file _default_ weights to be used for IPs that do not have a weight defined.

gabordozsa replied:
It makes sense, thanks for the suggestions! I changed the implementation to assign weight for every address record in the result list. If an address does not have a weight defined in the weight file then the default weight value "1" is assumed for it.  The answer is returned unaltered only if there isn't any address record.

The default weight value (1) is hardcoded currently as I'm not sure if it is really necessary to make that configurable (via the weight file).

chrisohaver commented on plugin/loadbalance/README.md:
```diff
@@ -19,7 +19,40 @@ implementations (like glibc) are particular about that.
+
+* **WEIGHTFILE** the weight file containing the weight values assigned to IPs. If the path is relative, the path from the **root** plugin will be prepended to it.
+
+The weight file is parsed line-by-line. If the first character in the line is '#' then the line is ignored (i.e. comment line). Otherwise, if the line contains a single word then it is a domain name. If the line contains two words then the first one is an IP and the second one is a weight value for that IP. The IPs are addresses for the domain name which is defined above them in the file. Multipl ...
```
I think this is probably describing verbatim the parsing algorithm... I think it would be clearer and shorter to just describe the expected syntax of the file, perhaps with an example file.

gabordozsa replied:
Okay, I will do that.

gabordozsa replied:
Done.

chrisohaver commented on plugin/loadbalance/README.md:
```diff
@@ -19,7 +19,40 @@ implementations (like glibc) are particular about that.
+The "round_robin" strategy randomizes the order of  A, AAAA, and MX records applying a uniform probability distribution.
+
+The "weighted_round_robin" policy assigns weight values to IPs to control the relative likelihood of particular IPs to be returned as top
+entry (A or AAAA record) in the answer. See [Wikipedia](https://en.wikipedia.org/wiki/Weighted_round_robin) about a generic
```
I think this reference is probably too generic and algorithmy to be helpful to someone who could not infer what weighted-round-robin is by its name.

gabordozsa replied:
Agree. I will remove it.

gabordozsa replied:
Done.

gabordozsa committed:
plugin/loadbalance: Add weighted round robin tests

Fix existing tests and add new ones for weighted round robin.

Signed-off-by: Gabor Dozsa <gabor.dozsa@ibm.com>

chrisohaver commented on plugin/loadbalance/weighted.go:
```diff
@@ -0,0 +1,299 @@
+	}
+)
+
+func (d *determininsticWRR) nextTopIP(weights []*weightItem, rn *rand.Rand) net.IP {
```
I'm not certain this is needed. While in theory, a deterministic implementation is different than random, IMO in the context of a server that typically serves a large number of requests, it is statistically/virtually identical to random.

gabordozsa replied:
I agree. I only added this for easier testing/demonstration of the behaviour (i.e. we can check the behaviour by running only a few queries instead of 100s or 1000s of queries). If you think that it is not necessary then I will remove it.

gabordozsa replied:
It is used in the unit tests, too, to verify expected results. (As opposed to the default round-robin policy which does not have any unit test to verify the expected order of the randomized result records.)

chrisohaver replied:
> It is used in the unit tests, too, to verify expected results.

Technically those tests are only testing the deterministic implementation.

gabordozsa replied:
True. I have removed the deterministic implementation.

chrisohaver commented on plugin/loadbalance/handler.go:
```diff
@@ -3,21 +3,54 @@ package loadbalance
 func (rr RoundRobin) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
-	wrr := &RoundRobinResponseWriter{w}
-	return plugin.NextOrFailure(rr.Name(), rr.Next, ctx, wrr, r)
+	rrw := &RoundRobinResponseWriter{ResponseWriter: w, policy: rr.policy, weights: rr.weights}
```
IIUC, the new weighted option only handles (or intends to handle) `A` and `AAAA` requests (see other note about SRV requests).  If so, we should not be using the RoundRobinResponseWriter when the query type is not `A`/`AAAA`, otherwise coredns could waste time looking through answers e.g. that won't have an A/AAAA records in them.

gabordozsa replied:
Right, I can change that.

But maybe I should change the entire weighted-round-robin approach. It could be an extra addition to the default round-robin policy in the sense that it would run default round-robin shuffle first always. After that, it would try to change the first item in the A/AAAA record lists (only) to honour the weights assigned to IPs. This way, it would handle the same requests as default round-robin does (with some extra tweaks for the A/AAAA records only). Would that be a better approach?

chrisohaver replied:
On the surface that would seem to create extra work when processing the result for no benefit?

gabordozsa replied:
Right, if the user is only interested in first A/AAAA record in the result list then it would not add any benefit. The only "benefit" would be a possible simpler implementation logic for the weighted-round-robin and more familiar results for users who are already familiar with the default round-robin. 

So I will stick with the current approach.  I'm going change the PR to handle only requests which have results with A/AAAA records in them.

gabordozsa replied:
Done.

chrisohaver commented on plugin/loadbalance/loadbalance.go:
```diff
@@ -18,9 +27,14 @@ func (r *RoundRobinResponseWriter) WriteMsg(res *dns.Msg) error {
+		res.Ns = roundRobin(res.Ns)
+		res.Extra = roundRobin(res.Extra)
+	case weightedRoundRobinPolicy:
+		res.Answer = r.weights.weightedRoundRobin(res.Question[0].Name, res.Answer)
```
I think the weighted shuffle implementation in its current state only shuffles answers to A/AAAA requests.  But if you want to shuffle responses to SRV requests too, for which any given SRV records can reference one _or more_ A/AAAA records in the `Extra` (aka Additional) section, you should also randomize Extra.  I think this would make the feature more complete - handling all (I think all) query types that can answer with A/AAAA records.

gabordozsa replied:
Right, I will add code to apply the logic to the Extra section, too.

gabordozsa replied:
Done.

chrisohaver reviewed (changes requested):
Thanks for contributing!  I've made some comments.

gabordozsa committed:
plugin/loadbalance: Limit query types for weighted-rr

Limit the weighted round robin policy to only handle A, AAAA and
SRV queries but apply the policy to the Extra section of the
answer, too.

Signed-off-by: Gabor Dozsa <gabor.dozsa@ibm.com>

gabordozsa committed:
plugin/loadbalance: Improve weight file description

Improve weight file description in the README. Add log message
about weight file updates.

Signed-off-by: Gabor Dozsa <gabor.dozsa@ibm.com>

gabordozsa committed:
plugin/loadbalance: Scrap determinstic weighted-rr

Scrap the deterministic implementation of weighted-round-robin.
(Keep only the randomized version.)

Signed-off-by: Gabor Dozsa <gabor.dozsa@ibm.com>

gabordozsa committed:
plugin/loadbalance: Use default weight value

Use default weight value 1 for any address record if weight is not
specified for it in the weight file. Only return result unaltered
if it contains no address record.

Signed-off-by: Gabor Dozsa <gabor.dozsa@ibm.com>

gabordozsa committed:
plugin/loadbalance: Fix lint errors

Fix lint and import format errors.

Signed-off-by: Gabor Dozsa <gabor.dozsa@ibm.com>

chrisohaver commented on plugin/loadbalance/README.md:
```diff
@@ -31,3 +76,29 @@ Load balance replies coming back from Google Public DNS:
+Use weighted round robin strategy to load balance replies defined by the **file** plugin:
+
+~~~ corefile
+demo.plmt {
```
Can we use `example.com` as the example zone?  I'm not sure what `plmt` and `plt` are referencing.

gabordozsa replied:
Sure, it's done.

gabordozsa replied:
Ack

chrisohaver commented on plugin/loadbalance/README.md:
```diff
@@ -19,7 +19,52 @@ implementations (like glibc) are particular about that.
+
+The "round_robin" strategy randomizes the order of  A, AAAA, and MX records applying a uniform probability distribution.
+
+The "weighted_round_robin" policy assigns weight values to IPs to control the relative likelihood of particular IPs to be returned as the first
```
I think just `weighted` should suffice for the name.

gabordozsa replied:
Done.

gabordozsa replied:
Yes, I agree. It's done.

gabordozsa committed:
plugin/loadbalance: Change weighted policy name

Change weighted_round_robin policy name to
weighted.

Signed-off-by: Gabor Dozsa <gabor.dozsa@ibm.com>

chrisohaver commented on plugin/loadbalance/loadbalance.go:
```diff
@@ -18,9 +27,20 @@ func (r *RoundRobinResponseWriter) WriteMsg(res *dns.Msg) error {
-	res.Answer = roundRobin(res.Answer)
-	res.Ns = roundRobin(res.Ns)
-	res.Extra = roundRobin(res.Extra)
+	switch r.policy {
```
Instead of policy being a string that is checked for every query, can you make it a function that is defined and set during setup? e.g. a func that operates on the dns message `func(*dns.Msg) *dns.Msg`

gabordozsa replied:
Yes, that is a nice improvement, thanks for the suggestion! It's done.

chrisohaver commented on plugin/loadbalance/handler.go:
```diff
@@ -3,21 +3,50 @@ package loadbalance
-}
+type (
+	// RoundRobin is a plugin to rewrite responses for "load balancing".
+	RoundRobin struct {
```
We should probably rename the plugin struct to something more generic.  `LoadBalance` seems more appropriate now.

gabordozsa replied:
Done.

chrisohaver commented on plugin/loadbalance/setup.go:
```diff
@@ -10,34 +13,102 @@ import (
 		return plugin.Error("loadbalance", err)
 	}
 
+	if policy == weightedRoundRobinPolicy {
```
Can you move this logic into `parse`, and have `parse` return onStartup/onShutdown functions instead of `weightedRR` ?

gabordozsa replied:
Done.

gabordozsa committed:
plugin/loadbalance: Avoid policy check at each query

Avoid policy check at each query by creating policy specific
shuffle functions during setup.

Signed-off-by: Gabor Dozsa <gabor.dozsa@ibm.com>

gabordozsa committed:
plugin/loadbalance: Rename the plugin struct

Rename the plugin struct to LoadBalance.

Signed-off-by: Gabor Dozsa <gabor.dozsa@ibm.com>

chrisohaver commented on plugin/loadbalance/handler.go:
```diff
@@ -10,15 +10,16 @@ import (
-func (rr RoundRobin) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
-	wrr := &RoundRobinResponseWriter{w}
-	return plugin.NextOrFailure(rr.Name(), rr.Next, ctx, wrr, r)
+func (rr LoadBalance) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
```
```suggestion
func (lb LoadBalance) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
```

gabordozsa replied:
Done.

chrisohaver commented on plugin/loadbalance/README.md:
```diff
@@ -19,7 +19,28 @@ implementations (like glibc) are particular about that.
+
+~~~
+loadbalance weighted_round_robin WEIGHTFILE
+~~~
```
Please remove this and consolidate it into with the first syntax definition (immediately under **Syntax** title).

gabordozsa replied:
Done.

chrisohaver commented on plugin/loadbalance/README.md:
```diff
@@ -19,7 +19,28 @@ implementations (like glibc) are particular about that.
+loadbalance weighted_round_robin WEIGHTFILE {
+			reload DURATION
+}
+~~~
```
Please remove this and consolidate it into with the first syntax definition (immediately under **Syntax** title).

gabordozsa replied:
Done.

chrisohaver commented on plugin/loadbalance/README.md:
```diff
@@ -31,3 +52,49 @@ Load balance replies coming back from Google Public DNS:
+
+This assigns weight vales `3`, `1` and `2` to the IPs `100.64.1.1`, `100.64.1.2` and `100.64.1.3`, respectively. These IPs are addresses in A records for the domain name `www.example.com`. (The IPs for the A records are defined in the `./db.example.com` file using the **file** plugin in this example).
+
+In this example, the ratio between the number of answers in which `100.64.1.1`, `100.64.1.2` or `100.64.1.3` is in the top (first) A record should converge to  `3 : 1 : 2`.  (E.g. there should be twice as many answers with `100.64.1.3` in the top A record than with `100.64.1.2`).
```
I think this description of what the example accomplishes should be stated before the example Corefile and weight file (i.e. moved to line 56 and maybe rephrased a bit).

gabordozsa replied:
Done.

chrisohaver reviewed (changes requested):
Thanks for being patient.  Ive added some comments on readme.

gabordozsa committed:
plugin/loadbalance: Fix imports ordering

Fix imports ordering.

Signed-off-by: Gabor Dozsa <gabor.dozsa@ibm.com>

gabordozsa committed:
plugin/loadbalance: Improve naming of variables

Improve naming of variables.

Signed-off-by: Gabor Dozsa <gabor.dozsa@ibm.com>

gabordozsa committed:
plugin/loadbalance: Fix option names in README

Fix option names in README.md.

Signed-off-by: Gabor Dozsa <gabor.dozsa@ibm.com>

gabordozsa committed:
plugin/loadbalance: Improve README

Fix policy names in README and re-arrange text.

Signed-off-by: Gabor Dozsa <gabor.dozsa@ibm.com>

gabordozsa committed:
plugin/loadbalance: Fix receiver name

Fix receiver name.

Signed-off-by: Gabor Dozsa <gabor.dozsa@ibm.com>

gabordozsa commented on plugin/loadbalance/handler.go:
```diff
@@ -10,15 +10,16 @@ import (
 
 // Name implements the Handler interface.
-func (rr RoundRobin) Name() string { return "loadbalance" }
+func (rr LoadBalance) Name() string { return "loadbalance" }
```
Done

chrisohaver commented on plugin/loadbalance/README.md:
```diff
@@ -31,3 +40,46 @@ Load balance replies coming back from Google Public DNS:
+
+where `ipXY` is an IP address for `domain-nameX` and `weightXY` is the weight value associated with that IP. The weight values are in the range of [1,255].
+
+The `weighted` policy selects one of the address record in the result list and moves it to the top (first) position in the list. The random selection takes into account the weight values assigned to the addresses in the weight file. If an address in the result list is associated with no weight value in the weight file then the default weight value "1" is assumed for it when the selection is perfo ...
```
Please move this general description of the weight file structure to a separate section `## Weightfile`, before  the Examples section.

gabordozsa replied:
Done.

gabordozsa committed:
plugin/loadbalance: Re-arrange README

Re-arrange README.

Signed-off-by: Gabor Dozsa <gabor.dozsa@ibm.com>

chrisohaver commented on plugin/loadbalance/README.md:
```diff
@@ -16,10 +16,43 @@ implementations (like glibc) are particular about that.
+(top) A/AAAA record in the answer. Note that it does not shuffle all the records in the answer, it is only concerned about the first A/AAAA record
+returned in the answer.
+
+ * **WEIGHTFILE** is the file containing the weight values assigned to IPs for various domain names. If the path is relative, the path from the **root** plugin will be prepended to it. The format is explained below in the *Examples* section.
```
```suggestion
 * **WEIGHTFILE** is the file containing the weight values assigned to IPs for various domain names. If the path is relative, the path from the **root** plugin will be prepended to it. The format is explained below in the *Weightfile* section.
```

gabordozsa replied:
Ops, I missed that one. Done.

gabordozsa committed:
plugin/loadbalance: Fix internal reference in README

Fix internal reference in README.

Signed-off-by: Gabor Dozsa <gabor.dozsa@ibm.com>
